libraries/networking/src/udt/Connection.cpp

```diff
@@ -0,0 +1,600 @@
+    
+    // reset the number of data packets received since last ACK
+    _packetsSinceACK = 0;
+    
+    _stats.record(ConnectionStats::Stats::SentACK);
+}
+
+void Connection::sendLightACK() {
+    SequenceNumber nextACKNumber = nextACK();
+    
+    if (nextACKNumber == _lastReceivedAcknowledgedACK) {
+        // we already got an ACK2 for this ACK we would be sending, don't bother
+        return;
+    }
+    
+    // create the light ACK packet, make it static so we can re-use it
+    static const int LIGHT_ACK_PACKET_PAYLOAD_BYTES = sizeof(SequenceNumber);
+    static auto lightACKPacket = ControlPacket::create(ControlPacket::ACK, LIGHT_ACK_PACKET_PAYLOAD_BYTES);
+    
+    // reset the lightACKPacket before we go to write the ACK to it
+    lightACKPacket->reset();
+    
+    // pack in the ACK
+    lightACKPacket->writePrimitive(nextACKNumber);
+    
+    // have the socket send off our packet immediately
+    _parentSocket->writeBasePacket(*lightACKPacket, _destination);
+    
+    _stats.record(ConnectionStats::Stats::SentLightACK);
+}
+
+void Connection::sendACK2(SequenceNumber currentACKSubSequenceNumber) {
+    // setup a static ACK2 packet we will re-use
+    static const int ACK2_PAYLOAD_BYTES = sizeof(SequenceNumber);
+    static auto ack2Packet = ControlPacket::create(ControlPacket::ACK2, ACK2_PAYLOAD_BYTES);
+    
+    // reset the ACK2 Packet before writing the sub-sequence number to it
+    ack2Packet->reset();
+    
+    // write the sub sequence number for this ACK2
+    ack2Packet->writePrimitive(currentACKSubSequenceNumber);
+    
+    // send the ACK2 packet
+    _parentSocket->writeBasePacket(*ack2Packet, _destination);
+    
+    // update the last sent ACK2 and the last ACK2 send time
+    _lastSentACK2 = currentACKSubSequenceNumber;
+    
+    _stats.record(ConnectionStats::Stats::SentACK2);
+}
+
+void Connection::sendNAK(SequenceNumber sequenceNumberRecieved) {
+    // create the loss report packet, make it static so we can re-use it
+    static const int NAK_PACKET_PAYLOAD_BYTES = 2 * sizeof(SequenceNumber);
+    static auto lossReport = ControlPacket::create(ControlPacket::NAK, NAK_PACKET_PAYLOAD_BYTES);
+    lossReport->reset(); // We need to reset it every time.
+    
+    // pack in the loss report
+    lossReport->writePrimitive(_lastReceivedSequenceNumber + 1);
+    if (_lastReceivedSequenceNumber + 1 != sequenceNumberRecieved - 1) {
+        lossReport->writePrimitive(sequenceNumberRecieved - 1);
+    }
+    
+    // have the parent socket send off our packet immediately
+    _parentSocket->writeBasePacket(*lossReport, _destination);
+    
+    // record our last NAK time
+    _lastNAKTime = high_resolution_clock::now();
+    
+    _stats.record(ConnectionStats::Stats::SentNAK);
+}
+
+void Connection::sendTimeoutNAK() {
+    if (_lossList.getLength() > 0) {
+        // construct a NAK packet that will hold all of the lost sequence numbers
```

Comment:
there's an assert in the packet creation, but we should probably force this to only add what can fit into one packet?